src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php

```diff
@@ -0,0 +1,158 @@
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @var CustomUrlRepository
+     */
+    private $customUrlRepository;
+
+    /**
+     * @var PathBuilder
+     */
+    private $pathBuilder;
+
+    public function __construct(
+        DocumentManagerInterface $documentManager,
+        CustomUrlRepository $customUrlRepository,
+        PathBuilder $pathBuilder
+    ) {
+        $this->documentManager = $documentManager;
+        $this->customUrlRepository = $customUrlRepository;
+        $this->pathBuilder = $pathBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create($webspaceKey, array $data, $locale = null)
+    {
+        $document = $this->documentManager->create('custom_urls');
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist(
+            $document,
+            $locale,
+            [
+                'parent_path' => $this->getItemsPath($webspaceKey),
+                'node_name' => Urlizer::urlize($document->getTitle()),
+                'load_ghost_content' => true,
+            ]
+        );
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function readList($webspaceKey)
+    {
+        // TODO pagination
+
+        return $this->customUrlRepository->findList($this->getItemsPath($webspaceKey));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($uuid, $locale = null)
+    {
+        return $this->documentManager->find($uuid, $locale, ['load_ghost_content' => true]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($uuid, array $data, $locale = null)
+    {
+        $document = $this->read($uuid, $locale);
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist($document, $locale);
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete($uuid)
+    {
+        $this->documentManager->remove($this->read($uuid));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFields()
+    {
+        return [
+            'title' => ['property' => 'title'],
+            'published' => ['property' => 'published'],
+            'baseDomain' => ['property' => 'baseDomain'],
+            'domainParts' => ['property' => 'domainParts', 'type' => 'json_array'],
+            'target' => ['property' => 'target', 'type' => 'reference'],
+            'multilingual' => ['property' => 'multilingual'],
+            'canonical' => ['property' => 'canonical'],
+            'redirect' => ['property' => 'redirect'],
+            'targetLocale' => ['property' => 'targetLocale'],
+        ];
+    }
+
+    /**
+     * Bind data array to given document.
+     *
+     * @param CustomUrlDocument $document
+     * @param array $data
+     * @param string $locale
+     */
+    private function bind(CustomUrlDocument $document, $data, $locale)
```

Comment:
I agree that this is not the optimal place, but just using forms here without putting enough thoughts into it will cause us even more trouble in the future. E.g. when introducing the csrf token for the login, the forms were affected, and saving pages was not working at all anymore. The form component does a lot more than simple mapping, so introducing another layer ourselves or using a library doing ONLY the mapping would of course be better, but that should be thought through.

In addition to that @alexander-schranz attended the Symfony Forms workshop, where @webmozart said that he wouldn't use the Forms for REST APIs. Maybe they can give us some more information on that.